server/isomorphic-routing/index.js

```diff
@@ -95,15 +94,27 @@ function compose( ...functions ) {
 	return functions.reduceRight( ( composed, f ) => () => f( composed ), () => {} );
 }
 
-export function getCacheKey( context ) {
-	if ( isEmpty( context.query ) || isEmpty( context.cacheQueryKeys ) ) {
-		return context.pathname;
+/**
+ * Get a key used to cache SSR result for the request, or null to disable the cache.
+ *
+ * @param  {Object}        context                The request context
+ * @param  {Array<string>} context.cacheQueryKeys Query parameter keys that should be cached
+ * @param  {string}        context.pathname       Request path
+ * @param  {Object}        context.query          Query parameters
+ * @return {?string}                              The cache key or null to disable cache
+ */
+export function getCacheKey( { cacheQueryKeys, pathname, query } ) {
+	// If we have a query, do not cache if any params are not in cacheQueryKeys
+	if ( ! isEmpty( query ) ) {
+		// Cache if our all of the query parameters are in cacheQueryKeys
+		if (
+			! isEmpty( cacheQueryKeys ) &&
+			Object.keys( query ).length === cacheQueryKeys.length &&
+			every( cacheQueryKeys, key => has( query, key ) )
+		) {
+			return pathname + '?' + deterministicStringify( query );
```

Comment:
This doesn't look like a URL any more, but all we need is something deterministic that we can serialize for use as a key.

You can see in the tests we'll now get strings like `/my/path?{"and_me":"2","cache_me":"1","me_too":"3"}`.

Reply:
What's the benefit of not keeping it a URL? I mean, why change it to that structure?

Reply:
https://github.com/Automattic/wp-calypso/pull/22593#issuecomment-366816978

`qs` is not stable, so requests to `/?a=1&b=2` and `/?b=2&a=1` could be cached differently. `sort`, which is intended to provide stable urls from `qs.stringify` that was used is not present in the version of `qs` we depend on, so had no impact.

This is a simple way to get a stable string from an object.

You can see the failing stability test here: https://circleci.com/gh/Automattic/wp-calypso/85034#tests/containers/1 

It was run on 73d4558, before this implementation was changed.

Reply:
> `sort`, which is intended to provide stable urls from `qs.stringify` that was used is not present in the version of `qs` we depend on, so had no impact.

I thought the reason might've perhaps been the notorious `localCompare` typo?

`a.localCompare( b )` should be `a.localeCompare( b )`

This was pointed out and [fixed](https://github.com/Automattic/wp-calypso/pull/19036/files#diff-d92d930a21711a960b300aeb231d9056R64) by @Tug in #19036, which had to be reverted, and we forgot to carry over the typo fix...

Reply:
😬 Could be, I just noticed that the test failed, and compared our version with the changelog:

> ## [**5.2.0**](https://github.com/ljharb/qs/issues?milestone=30&state=closed)
> - [**#64**](https://github.com/ljharb/qs/issues/64) Add option to sort object keys in the query string

https://github.com/ljharb/qs/blob/master/CHANGELOG.md#520

https://github.com/Automattic/wp-calypso/blob/eb1ed864c04599bde51b1c39a03ea74be553c14c/package.json#L123

Reply:
It looks like that string was wrong, but I did a quick verification by applying the below patch to 73d4558b030b236d3fe8ad67bc1a714d9e270e2e. Test still fails.

```patch
diff --git a/server/isomorphic-routing/index.js b/server/isomorphic-routing/index.js
index 68f309abfa..753e4487e4 100644
--- a/server/isomorphic-routing/index.js
+++ b/server/isomorphic-routing/index.js
@@ -115,7 +115,7 @@ export function getCacheKey( { cacheQueryKeys, pathname, query } ) {
 		) {
 			// Make a stable string representation
 			// @TODO: qs too old for sort param (added 5.2.0, fixed 6.1.0)
-			return pathname + '?' + qs.stringify( query, { sort: ( a, b ) => a.localCompare( b ) } );
+			return pathname + '?' + qs.stringify( query, { sort: ( a, b ) => a.localeCompare( b ) } );
 		}
 		return null;
 	}
```

Reply:
Let's go with `deterministicStringify` for now. I'll file a PR to bump the `qs` version so we can revisit and go back to normalized routes as cache keys later.

Reply:
> Let's go with deterministicStringify for now. I'll file a PR to bump the `qs` version

Sounds great, I'd be happy to get back to a url-like key 👍

Reply:
> I'll file a PR to bump the `qs` version

#23259

Reply:
>> I'll file a PR to bump the `qs` version
>
> #23259

Merged, FWIW 🙂